nn/transformer.py

```diff
@@ -0,0 +1,320 @@
+
+      self.num_layers = num_layers
+      self.norm = norm
+      self.norm_eps = norm_eps
+
+  def forward(self, x: nn.LayerRef, *, memory: nn.LayerRef, state: nn.LayerState) -> Tuple[nn.LayerRef, nn.LayerState]:
+    """
+    Applies every decoder layer initialized in self.layers.
+    """
+    if self.pos_enc is not None:
+      output = self.pos_enc(x)
+    else:
+      output = x
+
+    for mod in self.layers:
+      output, state = mod(output, memory=memory, state=state)
+
+    if self.norm is not None:
+      output = self.norm(output, epsilon=self.norm_eps)
+
+    return output, state
+
+  def initial_state(self) -> nn.LayerState:
+    """
+    initial state declaration
+    """
+    return nn.LayerState({mod: mod.initial_state() for mod in self.layers})
+
+
+class Transformer(nn.Module):
+  """
+  Standard Transformer Module
+  """
+  def __init__(self, output_dim: int = 512, num_heads: int = 8, num_encoder_layers: int = 6,
+               num_decoder_layers: int = 6, dim_ff: int = 2048, dropout: float = 0.1, att_dropout: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu,
+               custom_encoder: Optional[Any] = None, custom_decoder: Optional[Any] = None,
+               custom_encoder_layer: Optional[nn.LayerRef] = None, custom_decoder_layer: Optional[nn.LayerRef] = None,
+               norm_eps: float = 1e-5, norm=nn.layer_norm, encoder_pos_enc: Callable = nn.positional_encoding,
+               decoder_pos_enc: Callable = nn.positional_encoding, search: bool = False,
+               beam_size: Optional[int] = None) -> None:
+    """
+    :param output_dim: output dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param num_encoder_layers: Number of encoder layers
+    :param num_decoder_layers: Number of decoder layers
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param att_dropout: dropout value for attention
+    :param activation: activation function
+    :param custom_encoder: Custom Encoder to replace the standard encoder
+    :param custom_decoder: Custom Decoder to replace the standard decoder
+    :param custom_encoder_layer: Custom Encoder layer to replace the standard layer if custom_encoder and
+      custom_encoder_layer are given custom_encoder will be preferred
+    :param custom_decoder_layer: Custom Decoder layer to replace the standard layer if custom_decoder and
+      custom_decoder_layer are given custom_decoder will be preferred
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm: function for layer normalization
+    :param encoder_pos_enc: Positional Encoding for the Encoder
+    :param decoder_pos_enc: Positional Encoding for the Decoder
+    :param search: Whether search is enabled or not
+    :param beam_size: Beam Size for search
```

Comment:
beam_size and other search specific things should not be added here. This should somehow be more generic. Either via #18 or #49.